src/lib.rs

```diff
@@ -190,209 +224,268 @@ impl KafkaJsonToDelta {
-                        // buffer it.
-                        value_buffers.add(partition, m.offset(), value);
-                        self.log_message_buffered(&m, value_buffers.len()).await;
-                    } else if let Some(tx_version) = delta_writer
-                        .last_transaction_version(self.app_id.as_str())
-                        .await?
-                    {
-                        self.log_assignment_untracked_skipped(&m).await;
-
-                        // Since this is an untracked partition and a Delta transaction exists for
-                        // the table already, we need to update partition assignments, seek the
-                        // consumer and wait the next iteration of the run_loop.
-                        update_partition_assignment(
-                            self.partition_assignment.clone(),
-                            wal.clone(),
-                            Some(tx_version),
-                            &mut value_buffers,
-                            self.topic.as_str(),
-                            &self.consumer,
-                        )
-                        .await?;
-                        continue;
-                    } else {
-                        // Since there is no delta transaction - this should be the first message
-                        // available on the partition.
-                        // Update partition assignment and buffer.
-                        update_partition_assignment(
-                            self.partition_assignment.clone(),
-                            wal.clone(),
-                            None,
-                            &mut value_buffers,
-                            self.topic.as_str(),
-                            &self.consumer,
-                        )
-                        .await?;
+        Ok(())
+    }
 
-                        value_buffers.add(partition, m.offset(), value);
+    async fn deserialize_message(
+        &self,
+        msg: &BorrowedMessage<'_>,
+    ) -> Result<Value, ProcessingError> {
+        match deserialize_message(msg) {
+            Ok(v) => {
+                self.log_message_deserialized(msg).await;
+                Ok(v)
+            }
+            Err(KafkaJsonToDeltaError::KafkaMessageNoPayload { .. }) => {
+                self.log_message_deserialization_failed(msg).await;
+                Err(ProcessingError::Continue)
+            }
+            Err(KafkaJsonToDeltaError::KafkaMessageDeserialization { .. }) => {
+                self.log_message_deserialization_failed(msg).await;
+                Err(ProcessingError::Continue)
+            }
+            Err(e) => Err(e.into()),
+        }
+    }
 
-                        self.log_assignment_untracked_buffered(&m, value_buffers.len())
-                            .await;
-                    }
+    async fn transform_value(
+        &mut self,
```

Comment:
From @houqp's rec in https://delta-users.slack.com/archives/C01Q2RXCVSQ/p1620929864077200?thread_ts=1620929080.077000&cid=C01Q2RXCVSQ:

```suggestion
        &self,
```